File: assignments/assignment-1-solution/src/CrazyEights/Player/Hand.cs
using CrazyEights.Cards;
using CrazyEights.Domain;

namespace CrazyEights.Player;

/// <summary>
/// Represents a player's hand of cards.
/// </summary>
public class Hand
{
    private readonly List<ICard> _cards = new();

    public int CardCount => _cards.Count;

    public IReadOnlyList<ICard> CardList => _cards.AsReadOnly();

    public void AddCard(ICard card)
    {
        _cards.Add(card);
    }

    public void RemoveCard(ICard card)
    {
        if (!_cards.Contains(card))
            throw new InvalidOperationException("Card not found in hand.");
        _cards.Remove(card);
    }

    public IReadOnlyList<PlayableCard> GetPlayableCards(SuitType currentSuit, RankType currentRank)
    {
        return PlayableCardsSelector.Get(CardList, currentSuit, currentRank);
    }
}

File: assignments/assignment-1-solution/src-java/src/main/java/edu/kennesaw/crazy8s/player/Hand.java
package edu.kennesaw.crazy8s.player;

import edu.kennesaw.crazy8s.cards.Card;
import edu.kennesaw.crazy8s.domain.PlayableCard;
import edu.kennesaw.crazy8s.domain.PlayableCardsSelector;
import edu.kennesaw.crazy8s.domain.RankType;
import edu.kennesaw.crazy8s.domain.SuitType;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Represents a player's hand of cards.
 */
public class Hand {
    private final List<Card> cards = new ArrayList<>();

    public int getCardCount() {
        return cards.size();
    }

    public List<Card> getCardList() {
        return Collections.unmodifiableList(cards);
    }

    public void addCard(Card card) {
        cards.add(card);
    }

    public void removeCard(Card card) {
        if (!cards.remove(card)) {
            throw new IllegalStateException("Card not found in hand.");
        }
    }

    public List<PlayableCard> getPlayableCards(SuitType currentSuit, RankType currentRank) {
        return PlayableCardsSelector.get(getCardList(), currentSuit, currentRank);
    }
}
